nexus/src/db/queries/network_interface.rs

```diff
@@ -1521,22 +1663,108 @@ mod tests {
     use std::net::IpAddr;
     use std::sync::Arc;
     use uuid::Uuid;
 
+    // Add an instance. We'll use this to verify that the instance must be
+    // stopped to add or delete interfaces.
+    async fn create_instance(db_datastore: &DataStore) -> Instance {
+        let instance_id = Uuid::new_v4();
+        let project_id =
+            "f89892a0-58e0-60c8-a164-a82d0bd29ff4".parse().unwrap();
+        // Use the first chunk of the UUID as the name, to avoid conflicts.
+        // Start with a lower ascii character to satisfy the name constraints.
+        let name = format!("a{}", instance_id)[..9].parse().unwrap();
+        let params = InstanceCreate {
+            identity: IdentityMetadataCreateParams {
+                name,
+                description: "desc".to_string(),
+            },
+            ncpus: InstanceCpuCount(4),
+            memory: ByteCount::from_gibibytes_u32(4),
+            hostname: "inst".to_string(),
+            user_data: vec![],
+            network_interfaces: InstanceNetworkInterfaceAttachment::None,
+            disks: vec![],
+        };
+        let runtime = InstanceRuntimeState {
+            run_state: InstanceState::Creating,
+            sled_id: Uuid::new_v4(),
+            propolis_id: Uuid::new_v4(),
+            dst_propolis_id: None,
+            propolis_addr: Some(std::net::SocketAddr::new(
+                "::1".parse().unwrap(),
+                12400,
+            )),
+            migration_id: None,
+            hostname: params.hostname.clone(),
+            memory: params.memory,
+            ncpus: params.ncpus,
+            gen: Generation::new(),
+            time_updated: Utc::now(),
+        };
+        let instance =
+            Instance::new(instance_id, project_id, &params, runtime.into());
+        db_datastore
+            .project_create_instance(instance)
+            .await
+            .expect("Failed to create new instance record")
+    }
+
+    async fn create_stopped_instance(db_datastore: &DataStore) -> Instance {
+        let instance = create_instance(db_datastore).await;
+        instance_set_state(
+            db_datastore,
+            instance,
+            external::InstanceState::Stopped,
+        )
+        .await
+    }
+
+    async fn instance_set_state(
+        db_datastore: &DataStore,
+        mut instance: Instance,
+        state: external::InstanceState,
+    ) -> Instance {
+        let new_runtime = model::InstanceRuntimeState {
+            state: model::InstanceState::new(state),
+            gen: instance.runtime_state.gen.next().into(),
+            ..instance.runtime_state.clone()
+        };
+        let res = db_datastore
+            .instance_update_runtime(&instance.id(), &new_runtime)
+            .await;
+        assert!(matches!(res, Ok(true)), "Failed to stop instance");
+        instance.runtime_state = new_runtime;
+        instance
+    }
+
     #[tokio::test]
     async fn test_insert_network_interface_query() {
```

Comment:
We don't necessarily need to do this as a part of this PR, but I'd recommend splitting up this test into a handful of smaller ones.

Right now, this test is massive enough that if it fails, it could be for any number of reasons. We have multiple instances, a variety of VPCs and subnets (literally "subnet, "other_subnet", "other_valid_subnet"), we allocate instances in a loop, etc. Unfortunately, this also means that someone refactoring this test (possibly not you?) may not be aware of all the implicit behaviors you're trying to test.

In my experience, some of the best tests have the following form:

- They're named along the lines of: "test-**action**-causes-**result**"
- They *generally* try to be as single purpose as possible. If that's too expensive, even refactoring into different functions (but operation on the same DB) would be reasonable.

Looking at this test, I can eyeball a handful of tests that could easily be spun out:

- test_insert_nic_to_stopped_instance_ok
- test_insert_nic_to_running_instance_returns_bad_state
- test_insert_nic_to_missing_instance_returns_not_found
- test_insert_nic_automatic_ip_allocation_is_sequential
- test_insert_nic_ip_collision_returns_not_available
- test_insert_nic_name_collision_returns_invalid_request
- test_insert_nic_name_collision_different_instances_ok
- test_insert_mutiple_nics_in_different_subnets_ok
- test_insert_mutiple_nics_in_same_subnet_fails
- test_insert_mutiple_nics_in_different_vpc_fails
- test_insert_nic_ip_exhaustion

IMO seeing something like:

```
PASS - test_insert_nic_to_stopped_instance_ok
! FAIL ! - test_insert_nic_to_running_instance_returns_bad_state
! FAIL ! - test_insert_nic_to_missing_instance_returns_not_found
PASS - test_insert_nic_automatic_ip_allocation_is_sequential
PASS - test_insert_nic_ip_collision_returns_not_available
PASS - test_insert_nic_name_collision_returns_invalid_request
PASS - test_insert_nic_name_collision_different_instances_ok
PASS - test_insert_mutiple_nics_in_different_subnets_ok
PASS - test_insert_mutiple_nics_in_same_subnet_fails
PASS - test_insert_mutiple_nics_in_different_vpc_fails
PASS - test_insert_nic_ip_exhaustion
```
Would be much more informative (hey, looks like something is going wrong with checking instance state!) than seeing:

```
FAIL - test_insert_network_interface_query
```

Reply:
Yeah, I know I know :)

I've put that off indefinitely, because there's so much shared setup. But I'll factor that out into one or more functions, and split it up.